tortoise/README.md

```diff
@@ -0,0 +1,167 @@
+
+## Modes interaction
+
+Verifying mode is safer as tortoise doesn't have an imposed limit on the amount of counted votes (or atleast imposed limit can be signicantly higher). Naturally we want to stay in verifying mode as long as tortoise can make progress.
+
+#### When to switch into full mode?
+
+Tortoise certainly has to switch into full mode if there is a layer that wasn't finalized past hdist layers. In such case it needs to count previously uncounted ballots so that real margin can be compared with local threshold, see [voting](#voting). Without counting all votes we can get only pessimistic margin, which works for the purposes of the verifying tortoise, but may not work for self-healing purposes (e.g. instead of voting according to local threshold sign tortoise will vote according to weak coin).
+
+During rerun we expect verifying tortoise to make progress after counting `verifying-mode-verification-window` layers. This window can be as large as memory will allow us, computational complexity doesn't increase with the window. However if tortoise didn't make progress after counting votes within this window it will have to switch into full mode during rerun. 
+
+#### When to switch into verifying mode?
+
+It is safe to make an attempt at any point. However we need a sufficient number of ballots that vote consistently with each other in order for verifying tortoise to work. 
+
+In the implementation, we make an attempt after full tortoise made progress and changed local opinion on layers that were undecided according to the verifying tortoise. In the layer that was just finalized by full tortoise we filter ballots that ["can be good"](#good-ballots) with base ballots in the same state. After they are marked good we restart verifying tortoise and if same layer it made the same progress as full tortoise we switch the mode.
+
+| block/ballot | local opinion | 0xaa | 0xbb | 0xcc | 0xdd | 0xee | 0xff |
+| ------------ | ------------- | ---- | ---- | ---- | ---- | ---- | ---- |
+| base         | -             | -    | -    | 0xaa | 0xaa | 0xcc | 0xcc |
+| layer        | -             | 10   | 10   | 11   | 11   | 12   | 12   |
+| 0x11         | 1             | -1   | -1   | 1    | -1   | 1    | 1    |
+| 0x22         | 1             | -1   | -1   | 1    | -1   | 1    | 1    |
+| 0x33         | -1            | -    | -    | -1   | -1   | -1   | -1   |
+| 0x44         | -1            | -    | -    | -1   | -1   | -1   | -1   |
+| 0x55         | 1             | -    | -    | -    | -    | 1    | 1    |
+| 0x66         | -1            | -    | -    | -    | -    | -1   | -1   |
+
+In this example ballots starting from layer 10 are voting for some old layer (for example 9). Ballots from layer 10 disagree with out local opinion and will be marked bad. Ballot 0xcc from layer 10 and ballots 0xee and 0xff will be marked "can be good" because exceptions from those ballots are consistent with local opinion.
```

Comment:
```suggestion
In this example ballots starting from layer 10 are voting for some old layer (for example 9). Ballots from layer 10 disagree with our local opinion and will be marked bad. Ballot 0xcc from layer 10 and ballots 0xee and 0xff will be marked "can be good" because exceptions from those ballots are consistent with local opinion.
```